File: Leetcode/leetcode-0094-BinaryTreeInorderTraversal-[tree]-[easy]/Solution.recursive.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var inorderTraversal = function (root) {
    let result = [];
    if (root === null) {
        return result;
    }

    inorder(root, result);
    return result;
};

let inorder = (root, result) => {
    if (root === null) {
        return;
    }

    inorder(root.left, result);
    result.push(root.val);
    inorder(root.right, result);
};

File: Leetcode/leetcode-0144-BinaryTreePreorderTraversal-[tree]-[easy]/Solution.recursive.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        ArrayList<Integer> result = new ArrayList<>();
        preorder(root, result);
        return result;
    }
    
    public void preorder(TreeNode root, ArrayList<Integer> result) {
        if (root == null) {
            return;
        }
        
        result.add(new Integer(root.val));
        preorder(root.left, result);
        preorder(root.right, result);
    }
}